README.rst
This package exposes part of the xlocale-family of C functions to Python. This
is particularly useful to set the locale per-thread.


Normally you use the `locale.setlocale
<https://docs.python.org/3/library/locale.html#locale.setlocale>`_
function to change the locale for the entire process::

    import locale

    locale.setlocale(locale.LC_ALL, 'nl_NL')

This may not be desirable for webservers or other applications that use
multiple threads. To change the locale for just the current thread you
do this::

    import xlocale

    loc = xlocale.Locale(xlocale.LC_ALL_MASK, 'nl_NL')
    loc.use()


Creating a locale object
========================

`xlocale.Locale([mask[, locale[, base]])`
   Create a new locale object. The new locale is created by taken
   a base locale and changing one or more locale categories. If
   no base is given the C locale will be used.

   ``mask`` is a bitmask build from the ``LC_*_MASK`` constants. Normally you
   will want to use ``LC_ALL_MASK`` to change the locale completely.

   ``locale`` is a string with the locale name. If not provided the C locale
   will be used.

`xlocale.Locale.current_locale()`
   Return a Locale instance for the currently locale.


Locale instances
================

`Locale.use()`
  Switch the locale for the current thread.

`Locale.name(mask)`
  Return the name for a locale category. ``mask`` is a bitmask of locale
  categories, and the name for the first configured category will be returned.

  **This method is not available on all platforms, and will raise an exception
  on unsupported platforms.**

`Locale.lconv()`
  Return information describing how numbers and currency values must be
  displayed in the current locale. This is an object with the following
  attributes:

+--------------------+--------------------------------------------------------+
| decimal_point      | The decimal point character, except for currency       |
|                    | values, cannot be an empty string.                     |
+--------------------+--------------------------------------------------------+
| thousands_sep      | The separator between groups of digits before the      |
|                    | decimal point, except for currency values.             |
+--------------------+--------------------------------------------------------+
| grouping           | The sizes of the groups of digits, except for currency |
|                    | This is a list of integers representing group size     |
|                    | from low order digit groups to high order (right to    |
|                    | left). The list may be terminated with 0 or CHAR_MAX.  |
|                    | If the list is terminated with 0 the last group size   |
|                    | before the 0 is repeated to account for all the        |
|                    | digits. If the list isterminated with CHAR_MAX, no     |
|                    | more grouping  is performed.                           |
+--------------------+--------------------------------------------------------+
| int_curr_symbol    | The standardized international currency symbol.        |
+--------------------+--------------------------------------------------------+
| currency_symbol    | The local currency symbol.                             |
+--------------------+--------------------------------------------------------+
| mon_decimal_point  | The decimal point character for currency values.       |
+--------------------+--------------------------------------------------------+
| mon_thousands_sep  | The separator for digit groups in currency values.     |
+--------------------+--------------------------------------------------------+
| mon_grouping       | Like grouping but for currency values.                 |
+--------------------+--------------------------------------------------------+
| positive_sign      | The character used to denote nonnegative currency      |
|                    | values, usually the empty string.                      |
+--------------------+--------------------------------------------------------+
| negative_sign      | The character used to denote negative currency values, |
|                    | usually a minus sign.                                  |
+--------------------+--------------------------------------------------------+
| int_frac_digits    | The number of digits after the decimal point in an     |
|                    | international-style currency value.                    |
+--------------------+--------------------------------------------------------+
| frac_digits        | The number of digits after the decimal point in the    |
|                    | locale style for currency values.                      |
+--------------------+--------------------------------------------------------+
| p_cs_precedes      | True if the currency symbol precedes the currency      |
|                    | value for nonnegative values, False if it follows.     |
+--------------------+--------------------------------------------------------+
| p_sep_by_space     | True if a space is inserted between the currency       |
|                    | symbol and the currency value for nonnegative values,  |
|                    | False otherwise.                                       |
+--------------------+--------------------------------------------------------+
| n_cs_precedes      | Like p_cs_precedes but for negative values.            |
+--------------------+--------------------------------------------------------+
| n_sep_by_space     | Like p_sep_by_space but for negative values.           |
+--------------------+--------------------------------------------------------+
| p_sign_posn        | The location of the positive_sign with respect to a    |
|                    | nonnegative quantity and the currency_symbol, coded as |
|                    | follows:                                               |
|                    |                                                        |
|                    | * 0 - Parentheses around the entire string.            |
|                    | * 1 - Before the string.                               |
|                    | * 2 - After the string.                                |
|                    | * 3 - Just before currency_symbol.                     |
|                    | * 4 - Just after currency_symbol.                      |
+--------------------+--------------------------------------------------------+
| n_sign_posn        | Like p_sign_posn but for negative currency values.     |
+--------------------+--------------------------------------------------------+
| int_p_cs_precedes  | Same as p_cs_precedes, but for internationally         |
|                    | monetary quantities.                                   |
|                    | formatted monetary quantities.                         |
+--------------------+--------------------------------------------------------+
| int_n_cs_precedes  | Same as n_cs_precedes, but for internationally         |
|                    | monetary quantities.                                   |
|                    | formatted monetary quantities.                         |
+--------------------+--------------------------------------------------------+
| int_p_sep_by_space | Same as p_sep_by_space, but for internationally        |
|                    | formatted monetary quantities.                         |
+--------------------+--------------------------------------------------------+
| int_n_sep_by_space | Same as n_sep_by_space, but for internationally        |
|                    | formatted monetary quantities.                         |
+--------------------+--------------------------------------------------------+
| int_p_sign_posn    | Same as p_sign_posn, but for internationally formatted |
|                    | monetary quantities.                                   |
+--------------------+--------------------------------------------------------+
| int_n_sign_posn    | Same as n_sign_posn, but for internationally formatted |
|                    | monetary quantities.                                   |
+--------------------+--------------------------------------------------------+


Constants
=========

+----------------------+----------------------------------+
| Constant             | Description                      |
+======================+==================================+
| ``LC_COLLATE_MASK``  | Collation                        |
+----------------------+----------------------------------+
| ``LC_CTYPE_MASK``    | Character type                   |
+----------------------+----------------------------------+
| ``LC_MESSAGES_MASK`` | Messages                         |
+----------------------+----------------------------------+
| ``LC_MONETARY_MASK`` | Monetary                         |
+----------------------+----------------------------------+
| ``LC_NUMERIC_MASK``  | Numeric                          |
+----------------------+----------------------------------+
| ``LC_TIME_MASK``     | Time                             |
+----------------------+----------------------------------+
| ``LC_ALL_MASK``      | Combination of all of the above. |
+----------------------+----------------------------------+



changes.rst
Changelog
=========

1.3.4 - February 25, 2018
-------------------------

- More Python 3 compilation fixes.


1.3.3 - February 25, 2018
-------------------------

- Python 3 has no PyInt.


1.3.2 - February 25, 2018
-------------------------

- Fix handling of exceptions raised when creating a Locale: they are now
  correctly seen as being raised in the constructor.


1.3.1 - February 23, 2018
-------------------------

- Add Python 3 support.
- Add manylinux support.


1.2 - November 15, 2014
-----------------------

- No longer return the current locale from Locale.use(). This created
  unsolvable double free problems.

1.1 - October 10, 2014
----------------------

- Fix a memory handling error in Locale.use() which could result in a segfault
  when using a locale instance multiple times.


1.0 - August 27, 2014
---------------------

- First release



xlocale.c
#include <assert.h>
#ifdef PLATFORM_BSD
#include <xlocale.h>
#else
#include <locale.h>
#endif
#include <Python.h>
#include <structseq.h>
#include "localeconv.h"

static PyObject *LanguageError;
static PyTypeObject LocaleType;

#if PY_MAJOR_VERSION < 3
static PyMethodDef LanguageMethods[] = {
	{NULL, NULL, 0, NULL},
};
#endif

typedef struct {
	PyObject_HEAD
	locale_t locale;
} Locale;


static Locale* create_locale(PyTypeObject* type, locale_t loc) {
	Locale *self;
	if ((self=(Locale*)type->tp_alloc(type, 0))!=NULL)
		self->locale=loc;
	return self;
}


static PyObject* Locale_new(PyTypeObject* type, PyObject* args, PyObject* kwargs) {
	return (PyObject*)create_locale(type, LC_GLOBAL_LOCALE);
}


static void Locale_dealloc(Locale* self) {
	if (self->locale!=LC_GLOBAL_LOCALE)
		freelocale(self->locale);
	Py_TYPE(self)->tp_free((PyObject*)self);
}


static int Locale_init(Locale* self, PyObject* args, PyObject* kwargs) {
	static char *keywords[] = {"mask", "locale", "base", NULL}; 
	int mask = LC_ALL_MASK;
	const char* locale = NULL;
	PyObject* base = NULL;

	if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|izO!", keywords,
				&mask, &locale, &LocaleType, &base))
		return 0;

	locale_t loc = newlocale(mask, locale,
			base!=NULL ? ((Locale*)base)->locale : NULL);
	if (loc==NULL) {
		PyErr_SetString(PyExc_RuntimeError, "Error creating new Locale.");
		return -1;
	} else {
		if (self->locale!=LC_GLOBAL_LOCALE)
			freelocale(self->locale);
		self->locale=loc;
	}
	return 0;
}


static PyObject* Locale_current_locale(PyObject* cls) {
	locale_t loc = uselocale(NULL);
	if (loc==NULL) {
		PyErr_SetString(PyExc_RuntimeError, "Error getting current locale.");
		return NULL;
	}
	PyObject* result = (PyObject*)create_locale(&LocaleType, loc);
	if (result==NULL)
		freelocale(loc);
	return result;
}


static PyObject* Locale_use(Locale* self) {
	uselocale(self->locale);
	Py_RETURN_NONE;
}

static PyObject* Locale_name(Locale* self, PyObject* mask) {
#ifdef PLATFORM_BSD

#if PY_MAJOR_VERSION >= 3
	if (!PyLong_Check(mask)) {
#else
	if (!PyInt_Check(mask)) {
#endif
		PyErr_SetString(PyExc_ValueError, "Mask must be an integer.");
		return NULL;
	}

#if PY_MAJOR_VERSION >= 3
	int m = (int)PyLong_AsSsize_t(mask);
#else
	int m = (int)PyInt_AsSsize_t(mask);
#endif
	if (!m) {
		PyErr_SetString(PyExc_ValueError, "Mask may not be zero.");
		return NULL;
	}
	const char* name = querylocale(m, self->locale);
	if (name==NULL) {
		Py_RETURN_NONE;
	} else 
#if PY_MAJOR_VERSION >= 3
		return PyUnicode_FromString(name);
#else
		return PyString_FromString(name);
#endif

#else
	PyErr_SetString(PyExc_RuntimeError, "This OS does not support querylocalae()");
	return NULL;
#endif
}

#include "localeconv.c"


static PyMethodDef Locale_methods[] = {
	{"current_locale", (PyCFunction)Locale_current_locale,
		METH_NOARGS|METH_CLASS,
		"Get the currently active locale."},
	{"name", (PyCFunction)Locale_name, METH_O,
		"Get locale name for a specified category."},
	{"use", (PyCFunction)Locale_use, METH_NOARGS,
	 	"Switch the current thread to this locale."},
	{"lconv", (PyCFunction)Locale_localeconv, METH_NOARGS,
		"Return locale conventions"},
	{NULL},
};

static PyTypeObject LocaleType = {
	PyVarObject_HEAD_INIT(NULL, 0)
	"xlocale.Locale",		/* tp_name */ 
	sizeof(Locale),			/* tp_basicsize */ 
	0,				/* tp_itemsize */ 
	(destructor)Locale_dealloc,	/* tp_dealloc */ 
	0,				/* tp_print */ 
	0,				/* tp_getattr */ 
	0,				/* tp_setattr */ 
	0,				/* tp_compare */ 
	0,				/* tp_repr */ 
	0,				/* tp_as_number */ 
	0,				/* tp_as_sequence */ 
	0,				/* tp_as_mapping */ 
	0,				/* tp_hash  */ 
	0,				/* tp_call */ 
	0,				/* tp_str */ 
	0,				/* tp_getattro */ 
	0,				/* tp_setattro */ 
	0,				/* tp_as_buffer */ 
	Py_TPFLAGS_DEFAULT,		/* tp_flags */ 
	"Locale objects",		/* tp_doc */
	0,				/* tp_traverse */
	0,				/* tp_clear */
	0,				/* tp_richcompare */
	0,				/* tp_weaklistoffset */
	0,				/* tp_iter */
	0,				/* tp_iternext */
	Locale_methods,			/* tp_methods */
	0,				/* tp_members */
	0,				/* tp_getset */
	0,				/* tp_base */
	0,				/* tp_dict */
	0,				/* tp_descr_get */
	0,				/* tp_descr_set */
	0,				/* tp_dictoffset */
	(initproc)Locale_init,		/* tp_init */
	0,				/* tp_alloc */
	Locale_new,			/* tp_new */
};


void init_xlocale(PyObject* module) {
	PyModule_AddIntConstant(module, "LC_COLLATE_MASK", LC_COLLATE_MASK);
	PyModule_AddIntConstant(module, "LC_CTYPE_MASK", LC_CTYPE_MASK);
	PyModule_AddIntConstant(module, "LC_MESSAGES_MASK", LC_MESSAGES_MASK);
	PyModule_AddIntConstant(module, "LC_MONETARY_MASK", LC_MONETARY_MASK);
	PyModule_AddIntConstant(module, "LC_NUMERIC_MASK", LC_NUMERIC_MASK);
	PyModule_AddIntConstant(module, "LC_TIME_MASK", LC_TIME_MASK);
	PyModule_AddIntConstant(module, "LC_ALL_MASK", LC_ALL_MASK);

	LanguageError=PyErr_NewException("language.LanguageError", NULL, NULL);
	Py_INCREF(LanguageError);
	PyModule_AddObject(module, "LanguageError", LanguageError);

	if (PyType_Ready(&LocaleType)<0)
		return;
	Py_INCREF(&LocaleType);
	PyModule_AddObject(module, "Locale", (PyObject*)&LocaleType);
}


#if PY_MAJOR_VERSION >= 3
static struct PyModuleDef moduledef = {
	PyModuleDef_HEAD_INIT,
        "xlocale",	// m_name
        NULL,		// m_doc
        0,		// m_size
        NULL,		// m_methods
        NULL,		// m_slots
        NULL,		// m_traverse
        NULL,		// m_clear
        NULL		// m_free
};

#endif

PyMODINIT_FUNC
#if PY_MAJOR_VERSION >= 3
PyInit_xlocale(void) {
#else
initxlocale(void) {
#endif
	PyObject *module;

#if PY_MAJOR_VERSION >= 3
	module = PyModule_Create(&moduledef);
#else
	module = Py_InitModule("xlocale", LanguageMethods);
#endif

	if (module==NULL) {
#if PY_MAJOR_VERSION >= 3
		return NULL;
#else
		return;
#endif
	}

	init_xlocale(module);
	init_localconv(module);
#if PY_MAJOR_VERSION >= 3	
	return module;
#endif
}



localeconv.c
#ifndef PLATFORM_BSD
struct lconv* localeconv_l(locale_t loc) {
    locale_t old = uselocale(loc);
    struct lconv* result = localeconv();
    uselocale(old);
    return result;
}

size_t mbstowcs_l(wchar_t *wcstr, const char *mbstr, size_t count, locale_t loc) {
    locale_t old = uselocale(loc);
    size_t result = mbstowcs(wcstr, mbstr, count);
    uselocale(old);
    return result;
}
#endif


PyDoc_STRVAR(lconv__doc__,
		"lconv: Result from localconv");

static PyStructSequence_Field lconv_fields[] = {
/* 0 */	{"decimal_point",
		"The decimal point character, except for currency values, cannot be an empty string."},
/* 1 */	{"thousands_sep",
		"The separator between groups of digits before the decimal point, except\n"
		"for currency values."},
/* 2 */	{"grouping",
		"The sizes of the groups of digits, except for currency values.\n\n"
		"This is a list of integers representing group size from low order\n"
		"digit groups to high order (right to left). The list may be\n"
		"terminated with 0 or CHAR_MAX. If the list is terminated with 0,\n"
		"the last group size before the 0 is repeated to account for all\n"
		"the digits. If the list is terminated with CHAR_MAX,\n"
		"no more grouping is performed."},
/* 3 */	{"int_curr_symbol", "The standardized international currency symbol."},
/* 4 */	{"currency_symbol", "The local currency symbol."},
/* 5 */	{"mon_decimal_point", "The decimal point character for currency values."},
/* 6 */	{"mon_thousands_sep", "The separator for digit groups in currency values."},
/* 7 */	{"mon_grouping", "Like grouping but for currency values."},
/* 8 */	{"positive_sign",
		"The character used to denote nonnegative currency values, usually the\n"
		"empty string."},
/* 9 */	{"negative_sign",
		"The character used to denote negative currency values, usually a\n"
		"minus sign."},
/*10 */	{"int_frac_digits",
		"The number of digits after the decimal point in an international-\n"
		"style currency value."},
/*11 */	{"frac_digits",
		"The number of digits after the decimal point in the local style\n"
		"for currency values."},
/*12 */	{"p_cs_precedes",
		"True if the currency symbol precedes the currency value for nonnegative\n"
		"values, False if it follows."},
/*13 */	{"p_sep_by_space",
		"True if a space is inserted between the currency symbol and the currency\n"
		"value for nonnegative values, False otherwise."},
/*14 */	{"n_cs_precedes", "Like p_cs_precedes but for negative values."},
/*15 */	{"n_sep_by_space", "Like p_sep_by_space but for negative values."},
/*16 */	{"p_sign_posn",
		"The location of the positive_sign with respect to a nonnegative quantity\n"
		"and the currency_symbol, coded as follows:\n"
		"\n"
		"    0    Parentheses around the entire string.\n"
		"    1    Before the string.\n"
		"    2    After the string.\n"
		"    3    Just before currency_symbol.\n"
		"    4    Just after currency_symbol.\n"},
/*17 */	{"n_sign_posn", "Like p_sign_posn but for negative currency values."},
/*18 */	{"int_p_cs_precedes",
		"Same as p_cs_precedes, but for internationally formatted monetary\n"
		"quantities."},
/*19 */	{"int_n_cs_precedes",
		"Same as n_cs_precedes, but for internationally formatted monetary\n"
		"quantities."},
/*20 */	{"int_p_sep_by_space",
		"Same as p_sep_by_space, but for internationally formatted monetary\n"
		"quantities."},
/*21 */	{"int_n_sep_by_space",
		"Same as n_sep_by_space, but for internationally formatted monetary\n"
		"quantities."},
/*22 */	{"int_p_sign_posn",
		"Same as p_sign_posn, but for internationally formatted monetary\n"
		"quantities."},
/*23 */	{"int_n_sign_posn",
		"Same as n_sign_posn, but for internationally formatted monetary\n"
		"quantities."},
	{0},
};

static PyStructSequence_Desc lconv_desc = {
	"lconv",
	lconv__doc__,
	lconv_fields,
	24,
};

static PyTypeObject LconvType;


static void _seq_set_string(PyObject* seq, locale_t loc, ssize_t index, const char* data) {
	if (!data)
		return;
	size_t needed = mbstowcs_l(NULL, data, 0, loc);
	wchar_t space[30];
	wchar_t *buf;
	if (needed==(size_t)-1)
		// Should we log an error here?
		return;
	if (needed<sizeof(space))
		buf=space;
	else if ((buf=PyMem_Malloc(needed+1))==NULL)
		return;
	size_t length = mbstowcs_l(buf, data, needed+1, loc);
	PyStructSequence_SET_ITEM(seq, index, PyUnicode_FromWideChar(buf, length));
	if (buf!=space)
		PyMem_Free(buf);
}


static void _seq_set_bool(PyObject* seq, locale_t loc, ssize_t index, char data) {
	PyObject* item = PyBool_FromLong(data);

	if (item!=NULL)
		PyStructSequence_SET_ITEM(seq, index, item);
}


static void _seq_set_int(PyObject* seq, locale_t loc, ssize_t index, char data) {
#if PY_MAJOR_VERSION >= 3
	PyObject* item = PyLong_FromSize_t(data);
#else
	PyObject* item = PyInt_FromSize_t(data);
#endif
	if (item!=NULL)
		PyStructSequence_SET_ITEM(seq, index, item);
}


static void _seq_set_grouping(PyObject* seq, locale_t loc, ssize_t index, const char* data) {
	PyObject* groups = PyList_New(0);;

	if (groups==NULL)
		return;

	if (*data) {
		int i = -1;
		do {
			i++;

#if PY_MAJOR_VERSION >= 3
			PyObject* item = PyLong_FromLong(data[i]);
#else
			PyObject* item = PyInt_FromLong(data[i]);
#endif
			if (item==NULL)
				break;
			if (PyList_Append(groups, item)==-1) {
				Py_DECREF(item);
				Py_DECREF(groups);
				return;
			}
		} while (data[i] && data[i]!=CHAR_MAX);
	}
	PyStructSequence_SET_ITEM(seq, index, groups);
}


static PyObject* Locale_localeconv(Locale* self) {
	struct lconv* lc = localeconv_l(self->locale);
	int i = 0;
	PyObject* result = PyStructSequence_New(&LconvType);

	_seq_set_string(result, self->locale, i++, lc->decimal_point);
	_seq_set_string(result, self->locale, i++, lc->thousands_sep);
	_seq_set_grouping(result, self->locale, i++, lc->grouping);
	_seq_set_string(result, self->locale, i++, lc->int_curr_symbol);
	_seq_set_string(result, self->locale, i++, lc->currency_symbol);
	_seq_set_string(result, self->locale, i++, lc->mon_decimal_point);
	_seq_set_string(result, self->locale, i++, lc->mon_thousands_sep);
	_seq_set_grouping(result, self->locale, i++, lc->mon_grouping);
	_seq_set_string(result, self->locale, i++, lc->positive_sign);
	_seq_set_string(result, self->locale, i++, lc->negative_sign);
	_seq_set_int(result, self->locale, i++, lc->int_frac_digits);
	_seq_set_int(result, self->locale, i++, lc->frac_digits);
	_seq_set_bool(result, self->locale, i++, lc->p_cs_precedes);
	_seq_set_bool(result, self->locale, i++, lc->p_sep_by_space);
	_seq_set_int(result, self->locale, i++, lc->n_cs_precedes);
	_seq_set_int(result, self->locale, i++, lc->n_sep_by_space);
	_seq_set_int(result, self->locale, i++, lc->p_sign_posn);
	_seq_set_int(result, self->locale, i++, lc->n_sign_posn);
	_seq_set_bool(result, self->locale, i++, lc->int_p_cs_precedes);
	_seq_set_bool(result, self->locale, i++, lc->int_n_cs_precedes);
	_seq_set_bool(result, self->locale, i++, lc->int_p_sep_by_space);
	_seq_set_bool(result, self->locale, i++, lc->int_n_sep_by_space);
	_seq_set_int(result, self->locale, i++, lc->int_p_sign_posn);
	_seq_set_int(result, self->locale, i++, lc->int_n_sign_posn);
	assert(i==lconv_desc.n_in_sequence);

	return result;
}


void init_localconv(PyObject* module) {
	PyStructSequence_InitType(&LconvType, &lconv_desc);
	Py_INCREF((PyObject*)&LconvType);
	PyModule_AddObject(module, "lconv", (PyObject*)&LconvType);

}

